dts/arm/st: Add USB support for stm32l072/73

ydamigos opened the pull request:
Add USB support for stm32l072 and stm32l073 SoCs

Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>

ydamigos committed:
dts/arm/st: Add USB support for stm32l072/73

Add USB support for stm32l072 and stm32l073 SoCs

Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>

ydamigos committed:
drivers/usb/usb_dc_stm32: Provide EP_TYPE_* defines for L0

USB LL API provides the EP_TYPE_* defines. STM32Cube does not
provide USB LL API for STM32L0 family. Map EP_TYPE_* defines
to PCD_EP_TYPE_* defines.

Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>

tagunil commented on arch/arm/soc/st_stm32/stm32l0/dts.fixup:
```diff
@@ -42,4 +42,13 @@
 #define CONFIG_SPI_2_NAME                       ST_STM32_SPI_40003800_LABEL
 #define CONFIG_SPI_2_IRQ                        ST_STM32_SPI_40003800_IRQ_0
 
+
```
Do we need this empty line here?

tagunil commented:
@ydamigos It waits forever for HSI48 readiness. Will investigate.

tagunil commented:
Well, it works after enabling HSI48 buffer in SYSCFG:
```
--- a/drivers/usb/device/usb_dc_stm32.c
+++ b/drivers/usb/device/usb_dc_stm32.c
@@ -219,6 +219,10 @@ static int usb_dc_stm32_clock_enable(void)
         * STM32F030x4/x6/x8/xC and STM32F070x6/xB.
         */
 #if defined(RCC_HSI48_SUPPORT)
+#ifdef CONFIG_SOC_SERIES_STM32L0X
+       LL_SYSCFG_VREFINT_EnableHSI48();
+#endif /* CONFIG_SOC_SERIES_STM32L0X */
+
        LL_RCC_HSI48_Enable();
        while (!LL_RCC_HSI48_IsReady()) {
                /* Wait for HSI48 to become ready */
```
However, there is a problem: where should SYSCFG itself be enabled? I've made a quick hack to test USB, but I don't know where is the proper place to do it.

ydamigos commented:
>Well, it works after enabling HSI48 VREFINT buffer in SYSCFG.
>However, there is a problem: where should SYSCFG itself be enabled? I've made a quick hack to test USB, but I don't know where is the proper place to do it.

@tagunil Thanks, I will check it.

ydamigos committed:
drivers/usb_dc_stm32: HSI48 requires VREFINT in L0

In STM32L0 series, HSI48 requires VREFINT and its buffer
with 48 MHz RC to be enabled. This patch enables
VREFINT reference for HSI48 oscillator.

Signed-off-by: Ilya Tagunov <tagunil@gmail.com>
Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>

ydamigos commented:
@tagunil SYSCFG is enabled in soc-gpio.c file for stm32l0 series. I added a check which will raise an error if it is disabled.

tagunil commented:
@ydamigos It is enabled here only when some EXTI interrupt is configured. I think we should enable it unconditionally somewhere.

tagunil commented:
@ydamigos Looks good to me. It seems F0 needs this too, because of pin remapping for small packages.

ydamigos committed:
drivers/stm32l0x_ll_clock: Enable SYSCFG in clock_control

Enable System Configuration Controller clock in clock_control
instead of gpio controller when we set a gpio as interrupt.
We use System Configuration Controller to manage external
interrupts and internal voltage reference for 48 MHz HSI
used in USB device.

Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>

ydamigos commented:
@tagunil I applied the changes. Could you test it again? I would do the same for F0 family on a separate PR. My only problem is that I don't have the hw to test it.

tagunil commented:
@ydamigos How about enabling SYSGFG by direct LL calls as in F4/L4 instead of calling the same driver through generic clock control code?

ydamigos commented:
>How about enabling SYSGFG by direct LL calls as in F4/L4 instead of calling the same driver through generic clock control code?

I moved the code as it was. Where do we use in Zephyr's code LL calls to enable SYSCFG? I cannot find them.

tagunil commented:
> I moved the code as it was. Where do we use in Zephyr's code LL calls to enable SYSCFG? I cannot find them.
Not SYSCFG, but PWR is enabled by LL calls in stm32f4x_ll_clock.c and stm32l4x_ll_clock.c. I think SYSCFG could be enabled the similar way.

ydamigos commented:
>Not SYSCFG, but PWR is enabled by LL calls in stm32f4x_ll_clock.c (I was wrong about L4). I think SYSCFG could be enabled the similar way.

I made the change.

tagunil commented:
> I made the change.

There is a typo in function name: LL_APB2_GPR1_EnableClock should be LL_APB2_GRP1_EnableClock

tagunil commented:
After fixing the typo I've tried cdc_acm sample, it works just fine.

ydamigos commented:
>After fixing the typo I've tried cdc_acm sample, it works just fine.

I fixed the typo. Thanks for testing.

jfischer-no reviewed (commented):
LGTM
